numpy/random/bit_generator.pyi

```diff
@@ -0,0 +1,116 @@
+    Dict,
+    List,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from numpy import dtype, ndarray, uint32, uint64, unsignedinteger
+from numpy.typing import DTypeLike, _ArrayLikeInt_co, _DTypeLikeUInt, _ShapeLike, _SupportsDType
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+_T = TypeVar("_T")
+
+_UIntType = TypeVar("_UIntType", uint64, uint32)
+_DTypeLike = Union[
+    Type[_UIntType],
+    dtype[_UIntType],
+    _SupportsDType[dtype[_UIntType]],
+]
+
+class _Interface(NamedTuple):
+    state_address: Any
+    state: Any
+    next_uint64: Any
+    next_uint32: Any
+    next_double: Any
+    bit_generator: Any
+
+class ISeedSequence(abc.ABC):
+    @overload
+    @abc.abstractmethod
+    def generate_state(
+        self, n_words: int, dtype: _DTypeLike[_UIntType] = ...
+    ) -> ndarray[Any, dtype[_UIntType]]: ...
+    @overload
+    @abc.abstractmethod
+    def generate_state(
+        self, n_words: int, dtype: _DTypeLikeUInt = ...
+    ) -> ndarray[Any, dtype[unsignedinteger[Any]]]: ...
+
+class ISpawnableSeedSequence(ISeedSequence):
+    @abc.abstractmethod
+    def spawn(self: _T, n_children: int) -> List[_T]: ...
+
+class SeedlessSeedSequence(ISpawnableSeedSequence):
+    @overload
+    def generate_state(
+        self, n_words: int, dtype: _DTypeLike[_UIntType] = ...
+    ) -> ndarray[Any, dtype[_UIntType]]: ...
+    @overload
+    def generate_state(
+        self, n_words: int, dtype: _DTypeLikeUInt = ...
+    ) -> ndarray[Any, dtype[unsignedinteger[Any]]]: ...
+    def spawn(self: _T, n_children: int) -> List[_T]: ...
+
+class SeedSequence(ISpawnableSeedSequence):
+    entropy: Union[None, int, Sequence[int]]
+    spawn_key: Sequence[int]
+    pool_size: int
+    n_children_spawned: int
+    pool: ndarray[Any, dtype[uint32]]
+    def __init__(
+        self,
+        entropy: Union[None, int, Sequence[int]] = ...,
+        *,
+        spawn_key: Tuple[int, ...] = ...,
+        pool_size: int = ...,
+        n_children_spawned: int = ...,
+    ) -> None: ...
+    def __repr__(self) -> str: ...
+    @property
+    def state(
+        self,
+    ) -> Dict[str, Union[None, Sequence[int], int, Tuple[int, ...]]]: ...
+    def generate_state(self, n_words: int, dtype: DTypeLike = ...) -> ndarray[Any, Any]: ...
+    def spawn(self, n_children: int) -> List[SeedSequence]: ...
+
+class BitGenerator:
+    lock: Lock
+    def __init__(self, seed: Union[None, _ArrayLikeInt_co, SeedSequence] = ...) -> None: ...
+    def __getstate__(self) -> Dict[str, Any]: ...
+    def __setstate__(self, state: Dict[str, Any]) -> None: ...
+    def __reduce__(
+        self,
+    ) -> Tuple[Callable[[str], BitGenerator], Tuple[str], Tuple[Dict[str, Any]]]: ...
+    @property
+    def state(self) -> Mapping[str, Any]: ...
+    @state.setter
+    def state(self, value: Mapping[str, Any]) -> None: ...
```

Comment:
I'd actually suggest marking them as abstractmethods.
This is not quite true at runtime, but seems to be very much in the spirit of their implementation:
https://github.com/numpy/numpy/blob/7a18e4ac4f0b9f0933a190ac25cd75729bcdd146/numpy/random/bit_generator.pyx#L537-L554

Reply:
> spawn_key is still annotated as Tuple[int, ...] here (should be Sequence[int]).

While the external API accepts `Sequence[int]`, it is always stored on the attribute as a `Tuple[int,...]`.

> I'd actually suggest marking them as abstractmethods.

I've done this, and marked BitGenerator as `ABC`.  IIRC the only reason is isn't ABC is a limitation of Cython.

> Considering SeedSequence is not an abstractbaseclass I'd annotate either with a custom TypedDict or Dict[str, Any].

I have replaced with a TypedDict.